src/rpc/connection.rs

```diff
@@ -554,7 +554,7 @@ impl CircularQueue {
         rsp_meta_data: &mut [u8],
         rsp_data: &mut [u8],
     ) -> Result<u32, Box<dyn std::error::Error>> {
-        let id = self.end_index.fetch_add(1, Ordering::Relaxed);
+        let id = self.end_index.fetch_add(1, Ordering::Relaxed) % REQUEST_QUEUE_LENGTH as u32;
```

Comment:
The `end_index` is still not set to 0, right?

Reply:
even if end_index is still not set to 0, id could still be fetched with the expression above.

Reply:
this operation will not be atomic

Reply:
Should self.end_index be set `self.end_index.fetch_add(1, Ordering::Relaxed) % REQUEST_QUEUE_LENGTH as u32`?